README.md
tiq-server
==========

This is an HTTP API server for [tiq](https://github.com/imiric/tiq).

It's essentially an HTTP interface to interacting with `tiq` data stored in
PostgreSQL using [tiq-db](https://github.com/imiric/tiq-db).


Setup
-----

### Docker

Clone this repo and run:

```
$ make build
```

This will download and setup three [Docker](https://www.docker.io/) containers
with Ubuntu 14.04: one for PostgeSQL, one for nginx and one for tiq-server.

To start the servers run:
```
$ make
```
... which will make tiq-server (proxied behind nginx) available on
http://localhost:8000/.

If you wish to change the port, run:

```
$ make PORT=<number>
```

You can stop all containers with `make stop`.


### Manual

This is more complicated, but essentially `npm install -g tiq-server`, install
PostgreSQL (and optionally nginx), and configure and start everything manually.

Save yourself the trouble, and just use Docker. :)


Usage
-----

The API accepts and returns only JSON encoded data.

Here's how you add some tags to some text:

```
POST / HTTP/1.1
Content-Type: application/json

{"tokens":["This too shall pass."],"tags":["quotes","inspiring"]}
```

To tag some URLs:

```
POST / HTTP/1.1
Content-Type: application/json

{"tokens":["http://duckduckgo.com/"],tags":["url","search","awesome"]}
```

```
POST / HTTP/1.1
Content-Type: application/json

{"tokens":["http://www.bing.com/"],"tags":["url","search","microsoft"]}
```

```
POST / HTTP/1.1
Content-Type: application/json

{"tokens":["http://www.reddit.com/"],"tags":["url","timewaster","funny"]}
```

Then, to recall stuff tagged with `search`:
```
GET /?tags=search HTTP/1.1
```
->
```
{"status":"success","data":["http://duckduckgo.com/", "http://www.bing.com/"]}
```

... or both `search` and `awesome`:
```
GET /?tags=search&tags=awesome HTTP/1.1
```
->
```
{"status":"success","data":["http://duckduckgo.com/"]}
```

Or to get the tags associated with a token:
```
GET /?tags=http://duckduckgo.com/ HTTP/1.1
```
->
```
{"status":"success","data":["url","search","awesome"]}
```

Note that the string MAY be encoded (e.g. `/?tags=http%3A%2F%2Fduckduckgo.com%2F`),
which should return the same result.


You can also tag multiple things at once:
```
POST / HTTP/1.1
Content-Type: application/json

{"tokens":["http://www.engadget.com/","http://www.theverge.com/"],"tags":["url","tech"]}
```

Doing the reverse is equivalent to the above:
```
POST / HTTP/1.1
Content-Type: application/json

{"tags":["http://www.engadget.com/","http://www.theverge.com/"],"tokens":["url","tech"]}
```


By default, the `public` namespace is used if none is specified. But you can
use any other namespace you want to partition the data. For example:
```
POST /john HTTP/1.1
Content-Type: application/json

{"tokens":"http://myprivateblog.com/","tags":["url","blog"]}
```

Now to get things tagged with `url` under the `john` namespace:
```
GET /john?tags=url HTTP/1.1
```
->
```
{"status":"success","data":["http://myprivateblog.com/"]}
```


Configuration
-------------

The configuration file by default is expected in `$XDG_CONFIG_HOME/tiq-server/config.json`
(where `$XDG_CONFIG_HOME` is `$HOME/.config`).

Here are the options you can set in the configuration file:

- `port`: The port the server will listen on. [default: `8000`]
- `database.client`: The RDBMS client to use. One of `"sqlite3"`, `"pg"` or `"mysql"`.
  [default: `"pg"`]
- `database.connection.host`: Host name or IP address to connect to. [default: `"localhost"`]
- `database.connection.port`: Port to connect to. [default: `5432`]
- `database.connection.user`: Username used to connect to the host. [default: `null`]
- `database.connection.password`: Password used to connect to the host. [default: `null`]
- `database.connection.database`: Database name to use. [default: `"tiq"`]

Essentially, everything under `database` is what [tiq-db](https://github.com/imiric/tiq-db)
expects.

You can also change some of these at runtime using environment variables that
follow [Docker linking naming conventions](http://docs.docker.io/use/working_with_links_names/):

- `host`: `$DB_PORT_<NUM>_TCP_ADDR`
- `port`: `$DB_PORT_<NUM>_TCP_PORT`
- `user`: `$DB_ENV_USER`
- `password`: `$DB_ENV_PASSWORD`

License
-------

[MIT](LICENSE)


HISTORY.md
0.0.1 / 2014-05-18
==================

  * First public release.


test/index.js

/**
 * Module dependencies.
 */

var expectRequire = require('a').expectRequire,
    Promise = require('bluebird'),
    _ = require('lodash'),
    should = require('chai').should();


var describeStore = {
  singlens: ['test'],
  doublebleh: ['test with multiple tokens'],
  'http://example.com/': ['test with URL']
};

var associateArgs = null;

// Mock tiq-db
expectRequire('tiq-db').return(function() {
  return {
    describe: function(tokens, ns) {
      return Promise.resolve(describeStore[tokens.join('')+ns]);
    },
    associate: function() {
      associateArgs = _.values(arguments);
      return Promise.resolve();
    }
  }
});

var server = require('..');

describe('#associate()', function() {
  it('should associate tags with tokens using namespaces', function(done) {
    var options = {
      method: 'POST',
      url: '/private',
      payload: '{"tokens":["url"],"tags":["test","namespace"]}'
    };

    server.inject(options, function(response) {
      var result = response.result;
      response.statusCode.should.equal(200);
      result.status.should.equal('success');
      associateArgs.should.deep.equal([['url'], ['test', 'namespace'], 'private'])
      done();
    });
  });

  it('should support URLs in token', function(done) {
    var options = {
      method: 'POST',
      url: '/',
      payload: '{"tokens":["http://mycoolsite.com/"],"tags":["cool","test"]}'
    };

    server.inject(options, function(response) {
      var result = response.result;
      response.statusCode.should.equal(200);
      result.status.should.equal('success');
      associateArgs.should.deep.equal([['http://mycoolsite.com/'], ['cool', 'test'], ''])
      done();
    });
  });

  it('should return an error on malformed JSON', function(done) {
    var options = {
      method: 'POST',
      url: '/public',
      payload: '["test"'
    };

    server.inject(options, function(response) {
      var result = response.result;
      response.statusCode.should.equal(400);
      result.status.should.equal('error');
      done();
    });
  });
});

describe('#describe()', function() {
  it('should return the tags associated with the token using namespaces', function(done) {
    var options = {
      method: 'GET',
      url: '/ns?tags=single'
    };

    server.inject(options, function(response) {
      var result = response.result;
      response.statusCode.should.equal(200);
      result.status.should.equal('success');
      result.data.should.deep.equal(['test']);
      done();
    });
  });

  it('should return the tags associated with multiple tokens', function(done) {
    var options = {
      method: 'GET',
      url: '/?tags=double&tags=bleh'
    };

    server.inject(options, function(response) {
      var result = response.result;
      response.statusCode.should.equal(200);
      result.status.should.equal('success');
      result.data.should.deep.equal(['test with multiple tokens']);
      done();
    });
  });

  it('should support URLs in token', function(done) {
    var options = {
      method: 'GET',
      url: '/?tags=http://example.com/'
    };

    server.inject(options, function(response) {
      var result = response.result;
      response.statusCode.should.equal(200);
      result.status.should.equal('success');
      result.data.should.deep.equal(['test with URL'])
      done();
    });
  });

  it('should support encoded URLs in token', function(done) {
    var options = {
      method: 'GET',
      url: '/?tags=http%3A%2F%2Fexample.com%2F'
    };

    server.inject(options, function(response) {
      var result = response.result;
      response.statusCode.should.equal(200);
      result.status.should.equal('success');
      result.data.should.deep.equal(['test with URL'])
      done();
    });
  });
});


handlers.js

/**
 * Module dependencies.
 */
var Tiq = require('tiq-db'),
    _ = require('lodash'),
    tiq;

exports = module.exports = {
  init: function(config) {
    tiq = Tiq(config);
  },
  associate: associate,
  describe: describe
}

// Enum to standardize response status text
var Result = {
  SUCCESS: 'success',
  FAIL:    'fail',
  ERROR:   'error'
}

/**
 * Wraps the response data in a uniform object.
 */
function wrapResponse(result, data, extra) {
  return _.merge(
    {status: result},
    data ? {data: data} : {},
    extra || {}
  );
}


/**
 * Associate a collection of tokens with a collection of tags.
 */
function associate(request, reply) {
  var tokens = request.payload.tokens,
      tags = request.payload.tags,
      ns = request.params.namespace;

  tiq.associate(tokens, tags, ns).then(function(res) {
    reply(wrapResponse(Result.SUCCESS));
  });
}

/**
 * Get the tags associated with the given tokens.
 */
function describe(request, reply) {
  var tokens = request.query.tags,
      ns = request.params.namespace;

  // In case only a single one was passed
  if (typeof tokens === 'string') {
    tokens = [tokens];
  }

  tiq.describe(tokens, ns).then(function(tags) {
    reply(wrapResponse(Result.SUCCESS, tags));
  });
}


index.js
#!/usr/bin/env node

/**
 * Module dependencies.
 */
var path = require('path'),
    Hapi = require('hapi'),
    _ = require('lodash'),
    links = require('docker-links').parseLinks(process.env),
    handlers = require('./handlers');

function readJSON(file) {
  try {
    var data = require(file);
  } catch (err) {
    var data = {};
  }
  return data;
}

var defaultConfig = {
  port: 8000,
  database: {
    client: 'pg',
    connection: {
      host: (links.db && links.db.hostname) ? links.db.hostname : 'localhost',
      port: (links.db && links.db.port) ? links.db.port : 5432,
      user: process.env.DB_ENV_USER || null,
      password: process.env.DB_ENV_PASSWORD || null
    }
  }
};

var configFile = path.join(process.env.XDG_CONFIG_HOME ||
                    path.join(process.env.HOME, '.config'),
                    'tiq-server', 'config.json');

var config = _.extend(defaultConfig, readJSON(configFile));

// Setup the DB plugin
handlers.init(config.database);

var server = Hapi.createServer('0.0.0.0', config.port);

server.route([
  {
    method: 'GET',
    path: '/{namespace?}',
    handler: handlers.describe
  },
  {
    method: 'POST',
    path: '/{namespace?}',
    handler: handlers.associate
  },
]);

// Make error response payloads adhere to the schema used for non-errors
server.ext('onPreResponse', function (request, reply) {
  var response = request.response;
  if (!response.isBoom) {
    return reply();
  }

  delete response.output.payload.statusCode;
  delete response.output.payload.error;
  response.output.payload.status = 'error';

  reply(response);
});

if (require.main === module) {
  server.start();
}

module.exports = server;
